### PR TITLE
Add Edges and Edit Mode

### LINKED ISSUE #360: The C API (and the C++ API) do not seem to have the calls to addEdges
The C API has the calls MDAL_M_addVerices() and MDAL_M_addFaces() calls.

There does not seem to be an MDAL_M_addEdges() call and I cannot find the underlying C++ structure to do this.

This is despite at least one Driver advertising the ability to write Edge data.

Is there a fundamental reason behind this? 

If there is we should add it to the documentation.
If there is not then we should probably look at completing the API.

### LINKED ISSUE #365: The C API is not complete around Edit Modes
The handing of edit modes in the C API  needs to be improved:

- For a Mesh - it is created in Edit Mode but loaded not in Edit Mode. So the only way to edit a mesh is to create a new mesh. This IS workable and kind of makes sense because it avoids the complexities of data validation when editing, for instance, the vertices of a mesh with DataOnVertices etc. 

- The case for DataGroup is less clear. At the moment it is similar - created in Edit Mode but loaded read only. However, since read only prevents you from overwriting the complete dataset or adding a timeslice to the data (both perfectly reasonable actions) I think that this is not a sensible choice. The API should have the option to open Edit Mode as well as close Edit Mode.

I will submit a separate PR with proposed changes.

### PR DESCRIPTION
**Pull request by runette**
fixes #360 

This should be the last of updates needed for the next Python version.

- Adds the C and C++ methods to add edges to an editable mesh
- Updates and extends the API tests to include the edge API

### CODE REVIEW

File: mdal/api/mdal.h [hunk 1]
```diff
@@ -397,6 +413,8 @@ MDAL_EXPORT int MDAL_M_datasetGroupCount( MDAL_MeshH mesh );
 
 /**
  * Returns dataset group handle
+ *
+ * Since 0.9.0 - if the driver has Write Capability for this DatasetGroup, it is opened in edit mode
```
Comment (PeterPetrik) [inline mdal/api/mdal.h:417]: I strongly suggest to not change the behaviour of the exisisting functions. they are used all over the place in qgis and we would need to make sure to close edit mode? Also if you want to only read the dataset you cannot do it. Can you create a new function to start edit mode of existing mesh dataset group instead?

File: mdal/api/mdal.h [hunk 2]
```diff
@@ -647,6 +665,8 @@ MDAL_EXPORT MDAL_DatasetH MDAL_G_addDataset(
  * \param verticalExtrusion Double Array holding the vertical level values for the voxels
  *               Size must be Face count + Volume count
  * \returns empty pointer if not possible to create dataset (e.g. group opened in read mode), otherwise handle to new dataset
+ *
+ * \since MDAL 0.9.0
```
Comment (PeterPetrik) [inline mdal/api/mdal.h:669]: this is added in this version?

File: tests/test_api.cpp [hunk 3]
```diff
@@ -242,6 +242,91 @@ TEST( ApiTest, VerticesApi )
...
+                  );
+
+    EXPECT_TRUE( compareVectors( refStart, start ) );
+    EXPECT_TRUE( compareVectors( refStop, stop ) );
+  }
+  MDAL_CloseMesh( m );
+
+  // Some wrong calls tests
+  EXPECT_EQ( MDAL_M_edgeIterator( nullptr ), nullptr );
+  EXPECT_EQ( MDAL_EI_next( nullptr, 0, nullptr, nullptr ), 0 );
```
Comment (PeterPetrik) [inline tests/test_api.cpp:327]: should be separate test case and not mixed with regular tests

File: tests/test_api.cpp [hunk 4]
```diff
@@ -242,6 +242,91 @@ TEST( ApiTest, VerticesApi )
   EXPECT_EQ( MDAL_FI_next( nullptr, 0, nullptr, 0, nullptr ), 0 );
 }
 
+void _populateEdges( MDAL_MeshH m, std::vector<int> &start, std::vector<int> &stop, size_t itemsLen )
```
Comment (PeterPetrik) [inline tests/test_api.cpp:245]: static?

File: mdal/mdal_data_model.hpp [hunk 5]
```diff
@@ -279,6 +279,8 @@ namespace MDAL
 
       virtual void addVertices( size_t vertexCount, double *coordinates );
       virtual void addFaces( size_t faceCount, size_t driverMaxVerticesPerFace, int *faceSizes, int *vertexIndices );
+      virtual void addEdges( size_t edgeCount, int *startVertexIndices, int *endVertexIndices );
```
Comment (PeterPetrik) [inline mdal/mdal_data_model.hpp:282]: wondering if we shouldn't add new capability in https://github.com/lutraconsulting/MDAL/blob/5e8223ed283c15815089bcf7e5558e4b5660ac88/mdal/frmts/mdal_driver.hpp#L15 that driver is able to modify/write mesh frame (add faces/edges/vertices) ..? so at least we can know how we stand with driver support

**Review by PeterPetrik — COMMENTED** [review]
Thanks, I will have more time later this week to review your work!

File: mdal/api/mdal.h (hunk 1 above)
Comment (runette) [reply mdal/api/mdal.h:417]: Ok - I removed these changes from this PR and will add something else later

File: tests/test_api.cpp (hunk 3 above)
Comment (runette) [reply tests/test_api.cpp:327]: Not totally sure which part you are referring to and what you mean.

File: mdal/mdal_data_model.hpp (hunk 5 above)
Comment (runette) [reply mdal/mdal_data_model.hpp:282]: I do agree with you - see #361 - but I don't think that should be fixed in this PR and/or only for edges (which would create inconsistency) I suggest a separate dedicated PR to update the capabilities across all entities.

File: mdal/api/mdal.h (hunk 2 above)
Comment (runette) [reply mdal/api/mdal.h:669]: I think that is a copy/paste error. I will correct it.

Comment (runette) [reply mdal/api/mdal.h:669]: I have to correct myself - that is in the comments which are new in version 0.9.0 but in a previous PR. I just noticed that the version tag was not added in that PR

File: tests/test_api.cpp (hunk 3 above)
Comment (PeterPetrik) [reply tests/test_api.cpp:327]: just that I would not mix in on test function both tests that check that the api is working well with invalid inputs e.g. `EXPECT_EQ( MDAL_M_edgeIterator( nullptr ), nullptr );` and a normal test case where you test the api works for regular use case.

Comment (PeterPetrik) [reply tests/test_api.cpp:327]: but it is minor, do not worry

File: mdal/mdal_data_model.hpp (hunk 5 above)
Comment (PeterPetrik) [reply mdal/mdal_data_model.hpp:282]: ok

File: mdal/api/mdal.h (hunk 1 above)
Comment (PeterPetrik) [reply mdal/api/mdal.h:417]: can you also please change the api doc for this function accordingly?

Comment (PeterPetrik) [reply mdal/api/mdal.h:417]: @runette please change the api doc for this function accordingly? and i will merge afterwards

Comment (runette) [reply mdal/api/mdal.h]: API Doc corrected - good catch

**Review by PeterPetrik — APPROVED** [review]

**Pull request merged**